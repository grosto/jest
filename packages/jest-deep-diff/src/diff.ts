import getType = require('jest-get-type');
import {DiffObject, DiffPlugin, Memos, Path} from './types';
import {wrapIfHasCircularity} from './complex/circularObjects';
import {createEqual, createUnequalType, createUpdated} from './diffObject';
import {diffObjects} from './complex/object';
import {diffArrays} from './complex/array';
import {diffMaps} from './complex/map';
import {
  diffBooleans,
  diffDates,
  diffErrors,
  diffFunctions,
  diffNumbers,
  diffRegExps,
  diffStrings,
} from './primitives';

// eslint-disable-next-line @typescript-eslint/ban-types
type PrimitiveWrapper = String | Number | Boolean;

function diff(
  a: unknown,
  b: unknown,
  path?: Path,
  memos?: Memos,
  customDiffs: Array<DiffPlugin> = [],
): DiffObject {
  if (Object.is(a, b)) return createEqual(a, b, path);

  const diffWithCustomDiffs = (
    a: unknown,
    b: unknown,
    path?: Path,
    memos?: Memos,
  ) => diff(a, b, path, memos, customDiffs);

  for (const customDiff of customDiffs) {
    if (customDiff.test(a) || customDiff.test(b)) {
      return customDiff.diff(a, b, path, memos, diffWithCustomDiffs);
    }
  }

  const aType = getType(a);

  const bType = getType(b);
  if (aType !== bType) {
    return createUnequalType(a, b, path);
  }

  if (a instanceof Error) {
    return diffErrors(a, b as Error, path);
  }

  switch (aType) {
    case 'string':
      return diffStrings(a as string, b as string, path);
    case 'number':
      return diffNumbers(a as number, b as number, path);
    case 'boolean':
      return diffBooleans(a as boolean, b as boolean, path);
    case 'date':
      return diffDates(a as Date, b as Date, path);
    case 'regexp':
      return diffRegExps(a as RegExp, b as RegExp, path);
    case 'function':
      return diffFunctions(a as Function, b as Function, path);
  }

  // adapted from node assert
  // Use memos to handle cycles.
  if (memos === undefined) {
    memos = {
      a: new Map(),
      b: new Map(),
      position: 0,
    };
  } else {
    // We prevent up to two map.has(x) calls by directly retrieving the value
    // and checking for undefined. The map can only contain numbers, so it is
    // safe to check for undefined only.
    const bMemoA = memos.a.get(a);
    const bMemoB = memos.b.get(b);
    if (bMemoA !== undefined || bMemoB !== undefined) {
      return bMemoA === bMemoB
        ? createEqual(
            wrapIfHasCircularity(a, bMemoA),
            wrapIfHasCircularity(b, bMemoB),
            path,
          )
        : createUpdated(
            wrapIfHasCircularity(a, bMemoA),
            wrapIfHasCircularity(b, bMemoB),
            path,
          );
    }
    memos.position++;
  }

  memos.a.set(a, memos.position);
  memos.b.set(b, memos.position);

  if (!Object.is(Object.getPrototypeOf(a), Object.getPrototypeOf(b))) {
    return createUnequalType(a, b, path);
  }

  // TODO: DOM Node

  if (a instanceof Number || a instanceof String || a instanceof Boolean) {
    return Object.is(a.valueOf(), (b as PrimitiveWrapper).valueOf())
      ? createEqual(a, b, path)
      : createUpdated(a, b, path);
  }

  const diffWithMemosAndCustomDiffs = (a: unknown, b: unknown, path: Path) =>
    diffWithCustomDiffs(a, b, path, memos);

  switch (aType) {
    case 'array':
      return diffArrays(
        a as Array<unknown>,
        b as Array<unknown>,
        path,
        diffWithMemosAndCustomDiffs,
      );
    case 'object':
      return diffObjects(
        a as Record<string | number | symbol, unknown>,
        b as Record<string | number | symbol, unknown>,
        path,
        diffWithMemosAndCustomDiffs,
      );
    case 'map':
      return diffMaps(
        a as Map<unknown, unknown>,
        b as Map<unknown, unknown>,
        path,
        diffWithMemosAndCustomDiffs,
      );
    // case 'set':
    //   return compareSets(a as Set<unknown>, b as Set<unknown>, path, deepDiff);
    default:
      throw new Error('oopsie ' + aType);
  }
}

export default diff;
