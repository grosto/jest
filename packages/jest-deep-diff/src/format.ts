import chalk = require('chalk');
import {
  isKindDeleted,
  isKindEqual,
  isKindInserted,
  isKindUnequalType,
} from './diffObject';
import {
  DiffObject,
  Format,
  FormatOptions,
  FormatOptionsNormalized,
  Line,
  LineType,
} from './types';
import {formatCircularDiff, isWrappedCircular} from './complex/circularObjects';
import getType = require('jest-get-type');
import {normalizeDiffOptions} from './normalizeDiffOptions';
import {formatObjectDiff} from './complex/object';
import {formatMapDiff} from './complex/map';
import {formatArrayDiff} from './complex/array';
import {formatPrimitiveDiff, formatStringDiff} from './primitives';
import {createDeletedLine, createInsertedLine, formatUpdated} from './line';

export const printAnnotation = ({
  aAnnotation,
  aColor,
  aIndicator,
  bAnnotation,
  bColor,
  bIndicator,
  // includeChangeCounts,
  omitAnnotationLines,
}: FormatOptionsNormalized): string => {
  if (omitAnnotationLines) {
    return '';
  }

  const aRest = '';
  const bRest = '';

  return (
    aColor(aIndicator + ' ' + aAnnotation + aRest) +
    '\n' +
    bColor(bIndicator + ' ' + bAnnotation + bRest) +
    '\n\n'
  );
};

const addIntendation = (string: string, padding: string) =>
  string
    .split('\n')
    .map((x, i) => (i !== 0 ? '  ' : '') + padding + x)
    .join('\n');

function print(lines: Array<Line>, opts: FormatOptionsNormalized) {
  // eslint-disable-next-line consistent-return
  function printLine(line: Line) {
    const serializedVal = line.skipSerialize
      ? `${line.val}`
      : opts.serialize(line.val);

    const formattedLineContent =
      ' ' +
      addIntendation(line.prefix + serializedVal + line.suffix, line.indent);

    switch (line.type) {
      case LineType.COMMON:
        return opts.commonColor(opts.commonIndicator + formattedLineContent);
      case LineType.INSERTED:
        return opts.bColor(opts.bIndicator + formattedLineContent);
      case LineType.DELETED:
        return opts.aColor(opts.aIndicator + formattedLineContent);
    }
  }

  return lines.map(printLine).join('\n');
}

const formatDiff: Format = (diff, context, opts) => {
  for (const plugin of opts.plugins) {
    if (plugin.test(diff.a) || plugin.test(diff.b)) {
      return plugin.format(diff, context, opts, formatDiff);
    }
  }

  if (isKindUnequalType(diff.kind)) {
    return formatUpdated(diff.a, diff.b, context);
  }

  if (isKindInserted(diff.kind)) {
    return [createInsertedLine(diff.b, context)];
  }

  if (isKindDeleted(diff.kind)) {
    return [createDeletedLine(diff.a, context)];
  }

  if (isWrappedCircular(diff.a) || isWrappedCircular(diff.b)) {
    return formatCircularDiff(diff, context, opts);
  }

  const aType = getType(diff.a);
  switch (aType) {
    case 'number':
    case 'boolean':
    case 'function':
      return formatPrimitiveDiff(diff, context, opts);
    case 'string':
      return formatStringDiff(diff, context, opts);
    case 'array':
      return formatArrayDiff(
        diff as DiffObject<Array<unknown>>,
        context,
        opts,
        formatDiff,
      );
    case 'map':
      return formatMapDiff(
        diff as DiffObject<Map<unknown, unknown>>,
        context,
        opts,
        formatDiff,
      );
    case 'object':
      return formatObjectDiff(
        diff as DiffObject<Record<PropertyKey, unknown>>,
        context,
        opts,
        formatDiff,
      );
  }

  throw Error('ooppsie ');
};

function format(
  diff: DiffObject<unknown, unknown>,
  options: FormatOptions = {},
) {
  const normalizedOptions = normalizeDiffOptions(options);
  if (isKindEqual(diff.kind)) {
    return chalk.dim('Compared values have no visual difference.');
  }

  if (isKindUnequalType(diff.kind)) {
    return (
      '  Comparing two different types of values.' +
      ` Expected ${chalk.green(getType(diff.a))} but ` +
      `received ${chalk.red(getType(diff.b))}.`
    );
  }

  const header = printAnnotation(normalizedOptions);

  return (
    header +
    print(
      formatDiff(
        diff,
        {indent: '', prefix: '', sufix: ''},
        {plugins: normalizedOptions.plugins},
      ),
      normalizedOptions,
    )
  );
}

export default format;
