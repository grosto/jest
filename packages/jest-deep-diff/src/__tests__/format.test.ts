import stripAnsi = require('strip-ansi');
import {alignedAnsiStyleSerializer} from '@jest/test-utils';
import format from '../format';
import diff from '../diff';

expect.addSnapshotSerializer(alignedAnsiStyleSerializer);

const strippedFormat = valueDiff =>
  stripAnsi(format(valueDiff, {omitAnnotationLines: true}));

test('numbers', () => {
  expect(strippedFormat(diff(1, 2))).toBe('- 1\n+ 2');
});

test('-0 and 0', () => {
  expect(strippedFormat(diff(-0, 0))).toBe('- -0\n+ 0');
});

test('booleans', () => {
  expect(strippedFormat(diff(false, true))).toBe('- false\n+ true');
});

test('online strings', () => {
  expect(strippedFormat(diff('banana', 'apple'))).toBe('- banana\n+ apple');
});

describe('Objects', () => {
  it('should format object updated', () => {
    const a = {a: {b: {c: 5, d: 6}}};
    const b = {a: {b: {c: 6, d: 6}}};

    const expected = [
      '  Object {',
      '    "a": Object {',
      '      "b": Object {',
      '-       "c": 5,',
      '+       "c": 6,',
      '        "d": 6,',
      '      },',
      '    },',
      '  }',
    ].join('\n');

    expect(strippedFormat(diff(a, b))).toEqual(expected);
  });

  test('if property is set to undefined it should show up in diff', () => {
    const a = {a: 2};
    const b = {a: undefined};
    const expected = [
      '  Object {',
      '-   "a": 2,',
      '+   "a": undefined,',
      '  }',
    ].join('\n');

    expect(strippedFormat(diff(a, b))).toEqual(expected);
  });

  test('one property is complex and other is primitive', () => {
    const a = {a: 2};
    const b = {
      a: {
        b: {
          c: 3,
        },
      },
    };
    const expected = [
      '  Object {',
      '-   "a": 2,',
      '+   "a": Object {',
      '      "b": Object {',
      '        "c": 3,',
      '      },',
      '    },',
      '  }',
    ].join('\n');
    expect(strippedFormat(diff(a, b))).toEqual(expected);
  });

  describe.skip('multiline string as value of object property', () => {
    const expected = [
      '  Object {',
      '    "id": "J",',
      '    "points": "0.5,0.460',
      '+ 0.5,0.875',
      '  0.25,0.875",',
      '  }',
    ].join('\n');

    test('(non-snapshot)', () => {
      const a = {
        id: 'J',
        points: '0.5,0.460\n0.25,0.875',
      };
      const b = {
        id: 'J',
        points: '0.5,0.460\n0.5,0.460\n0.25,0.875',
      };

      const actual = strippedFormat(diff(a, b));
      console.log(actual);
      expect(actual).toBe(expected);
    });
  });
});

describe('multiline string non-snapshot', () => {
  // For example, CLI output
  // toBe or toEqual for a string isnâ€™t enclosed in double quotes.
  const a = `Options:
--help, -h  Show help                            [boolean]
--bail, -b  Exit the test suite immediately upon the first
            failing test.                        [boolean]`;
  const b = `Options:
  --help, -h  Show help                            [boolean]
  --bail, -b  Exit the test suite immediately upon the first
              failing test.                        [boolean]`;
  const expected = [
    '  Options:',
    '- --help, -h  Show help                            [boolean]',
    '- --bail, -b  Exit the test suite immediately upon the first',
    '-             failing test.                        [boolean]',
    '+   --help, -h  Show help                            [boolean]',
    '+   --bail, -b  Exit the test suite immediately upon the first',
    '+               failing test.                        [boolean]',
  ].join('\n');

  test('(unexpanded)', () => {
    expect(strippedFormat(diff(a, b))).toBe(expected);
  });
  test('(expanded)', () => {
    expect(strippedFormat(diff(a, b))).toBe(expected);
  });
});

describe('Arrays', () => {
  test('(unexpanded)', () => {
    const a = [1, 4, 4];
    const b = [1, 6];

    const expected = [
      '  Array [',
      '    1,',
      '-   4,',
      '+   6,',
      '-   4,',
      '  ]',
    ].join('\n');
    const actual = strippedFormat(diff(a, b));
    expect(actual).toBe(expected);
  });

  test('nested objects', () => {
    const a = [1, {a: 3, b: 2, c: 20}, 4];
    const b = [1, {a: 1, b: 2, d: 30}];
    const actual = strippedFormat(diff(a, b));
    const expected = [
      '  Array [',
      '    1,',
      '    Object {',
      '-     "a": 3,',
      '+     "a": 1,',
      '      "b": 2,',
      '-     "c": 20,',
      '+     "d": 30,',
      '    },',
      '-   4,',
      '  ]',
    ].join('\n');

    expect(actual).toBe(expected);
  });

  test('nested arrays', () => {
    const a = [1, [1, 3, 3], 4];
    const b = [1, [1, 4]];
    const actual = strippedFormat(diff(a, b));
    const expected = [
      '  Array [',
      '    1,',
      '    Array [',
      '      1,',
      '-     3,',
      '+     4,',
      '-     3,',
      '    ],',
      '-   4,',
      '  ]',
    ].join('\n');

    expect(actual).toBe(expected);
  });
});

describe.skip('Maps', () => {
  test('primitive keys', () => {
    const a = new Map([
      ['a', 1],
      ['b', 3],
      ['c', 3],
    ]);
    const b = new Map([
      ['a', 1],
      ['b', 2],
      ['d', 4],
    ]);

    const actual = strippedFormat(diff(a, b));
    const expected = [
      '  Map {',
      '    "a" => 1,',
      '-   "b" => 3,',
      '+   "b" => 2,',
      '-   "c" => 3,',
      '+   "d" => 4,',
      '  }',
    ].join('\n');

    expect(actual).toEqual(expected);
  });

  test('complex keys', () => {
    const a = new Map([
      [{a: 1}, {a: 2}],
      [{a: 2}, {a: 3}],
    ]);
    const b = new Map([
      [{a: 3}, {a: 2}],
      [{b: 2}, {a: 3}],
    ]);

    const expected = [
      '  Map {',
      '    Object {',
      '-     "a": 1,',
      '+     "a": 3,',
      '    } => Object {',
      '      "a": 2,',
      '    },',
      '    Object {',
      '-     "a": 2,',
      '+     "b": 2,',
      '    } => Object {',
      '      "a": 3,',
      '    },',
      '  }',
    ].join('\n');

    const actual = strippedFormat(diff(a, b));
    expect(actual).toEqual(expected);
  });
});

describe('Sets', () => {
  test.todo('marks updated values');
});

describe('multiline strings', () => {
  const a = `line 1
line 2
line 3
line 4`;
  const b = `line 1
line  2
line 3
line 4`;
  const expected = [
    '  line 1',
    '- line 2',
    '+ line  2',
    '  line 3',
    '  line 4',
  ].join('\n');

  test('(unexpanded)', () => {
    expect(strippedFormat(diff(a, b))).toBe(expected);
  });
});

describe('circular values', () => {
  test('(unexpanded)', () => {
    const a: Record<string, unknown> = {};
    a.x = {y: a};
    const b: Record<string, unknown> = {};
    const bx: Record<string, unknown> = {};
    b.x = bx;
    bx.y = bx;
    expect(strippedFormat(diff(a, b))).toMatchSnapshot();
  });

  test.skip('nested circular values', () => {
    const a: Record<string, unknown> = {b: 1};
    a.x = {y: a};
    const b: Record<string, unknown> = {};
    b.x = {y: b};

    expect(strippedFormat(diff(a, b))).toMatchSnapshot();
  });

  test('only one side has circularity', () => {
    const a: Record<string, unknown> = {};
    a.x = {y: a};
    const b = {
      x: {
        y: {
          a: {
            b: 3,
          },
        },
      },
    };

    expect(strippedFormat(diff(a, b))).toMatchSnapshot();

    expect(strippedFormat(diff(b, a))).toMatchSnapshot();
  });
});
