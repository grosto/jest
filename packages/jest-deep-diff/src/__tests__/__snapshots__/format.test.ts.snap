// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`circular values (unexpanded) 1`] = `
  Object {
    "x": Object {
-     "y": [Circular to depth 0],
+     "y": [Circular to depth 1],
    },
  }
`;

exports[`circular values nested circular values 1`] = `
  Object {
-   "b": 1,
    "x": Object {
      "y": [Circular to depth 0],
    },
  }
`;

exports[`circular values only one side has circularity 1`] = `
  Object {
    "x": Object {
-     "y": [Circular to depth 0],
+     "y": Object {
        "a": Object {
          "b": 3,
        },
      },
    },
  }
`;

exports[`circular values only one side has circularity 2`] = `
  Object {
    "x": Object {
-     "y": Object {
        "a": Object {
          "b": 3,
        },
      },
+     "y": [Circular to depth 0],
    },
  }
`;
